hal/src/nRF52840/ble_hal.cpp

```diff
@@ -1458,16 +1465,10 @@ int BleObject::Observer::stopScanning() {
     if (sd_ble_gap_scan_stop() != NRF_SUCCESS) {
         // LOG(ERROR, "Device is not in scanning state.");
     }
-    bool give = false;
-    ATOMIC_BLOCK() {
```

Comment:
We might give the semaphore in BLE event thread on `BLE_GAP_EVT_TIMEOUT` event, followed by giving the semaphore herein again, it may result in simple return on next call to `startScanning()` when it tries taking the semaphore.

Reply:
I think this should be fine as long as we make sure to guard `isScanning_` and further steps in `stopScanning()` under a BLE lock (same as it is right now on Gen 4). Is it safe to acquire it from inside the event handler for `BLE_GAP_EVT_TIMEOUT` ? You are more familiar with this.

Reply:
Currently the `BLE_GAP_EVT_TIMEOUT` event is handled in ISR directly, unlike other events those are push to the event queue. We can do that for the `BLE_GAP_EVT_TIMEOUT` event and acquire the lock in the event thread, but we probably run into deadlock. For instance,
```
Call a BLE API -> acquire lock -> waiting for event X to give semaphore -> release lock
                                  |-> handle event X in event thread -> waiting for the lock to releae
```
In that case we need to release the lock after the SDK API returns as same as in `startScanning()` for most of the BLE HAL APIs. This seams to be a significant change and require a ton of test work to make sure it won't introduce any regression. 
So the existing logic is right guarding the `isScanning_`, instead of using the lock. I would leave it as-is for now.

Reply:
Yeah, I double-checked with how it is right now, I think it should work fine as-is, so let's not overcomplicate this.